Update ReflectionFunctionAbstract.php

lanlin opened the pull request:
Fix The Error : Typed static property Roave\BetterReflection\Reflection\ReflectionFunctionAbstract::$parser must not be accessed before initialization

lanlin committed:
Update ReflectionFunctionAbstract.php

Fix The Error : Typed static property Roave\BetterReflection\Reflection\ReflectionFunctionAbstract::$parser must not be accessed before initialization

Ocramius commented:
Access to this property is indeed not supposed to happen upfront: how can
we reproduce the crash?

Can you write a failing test case please?

On Tue, Jun 22, 2021, 10:02 兰林 ***@***.***> wrote:

> Fix The Error : Typed static property
> Roave\BetterReflection\Reflection\ReflectionFunctionAbstract::$parser must
> not be accessed before initialization
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Roave/BetterReflection/pull/746
> Commit Summary
>
>    - Update ReflectionFunctionAbstract.php
>
> File Changes
>
>    - *M* src/Reflection/ReflectionFunctionAbstract.php
>    <https://github.com/Roave/BetterReflection/pull/746/files#diff-04bcae2cbb77965693175d351f3c4d8ce269f6cde240127a6f1617b43812cac6>
>    (2)
>
> Patch Links:
>
>    - https://github.com/Roave/BetterReflection/pull/746.patch
>    - https://github.com/Roave/BetterReflection/pull/746.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Roave/BetterReflection/pull/746>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABFVEEHC35VRQRSBSUZIRTTUA7RPANCNFSM47DEKQ2A>
> .
>

lanlin commented:
It takes me much time to reproduce this error

PHP Version: 8.0.1
PHPUnit Version: 9.5

```php
require 'vendor/autoload.php';

use PHPUnit\Framework\TestCase;
use Roave\BetterReflection\Reflection\ReflectionFunctionAbstract;

class Tests extends TestCase
{
    // ------------------------------------------------------------------------------

    public function testA(): void
    {
        $a = ReflectionFunctionAbstract::CLOSURE_NAME;
    }

    // ------------------------------------------------------------------------------

    /**
     * @dataProvider getData
     */
    public function testB(array $data): void
    {
        $b = ReflectionFunctionAbstract::CLOSURE_NAME;
    }

    // ------------------------------------------------------------------------------

    /**
     * @return array[]
     */
    public function getData(): array
    {
        return [
            [['aaa']],
            [['bbb']],
        ];
    }

    // ------------------------------------------------------------------------------
}
```

when using with PHPUnit's `@dataProvider` this error occurred.

Ocramius commented:
> It takes me much time to reproduce this error

It takes ages to maintain libraries without test cases :-) 

Could you factor the above in a test in either https://github.com/Roave/BetterReflection/blob/86135105e7f6e9684efd65c688ab2c49cf240143/test/unit/Reflection/ReflectionFunctionAbstractTest.php#L44 or any of the concrete types around it?

lanlin commented:
@Ocramius ok, give me some time

lanlin committed:
add unit test case for the property $parser

add unit test case for the property $parser
"Error : Typed static property Roave\BetterReflection\Reflection\ReflectionFunctionAbstract::$parser must not be accessed before initialization"

Ocramius commented on test/unit/Reflection/ReflectionFunctionAbstractTest.php:
```diff
@@ -745,6 +745,32 @@ function foo() {}', $comment);
+    {
+        $x = ReflectionFunctionAbstract::CLOSURE_NAME;
+
+        $this->assertEquals('{closure}', $x);
```
This is verifying a constant value is equal something, and it does not use `$data`? :thinking: 

Is this a WIP test?

lanlin replied:
it's nothing about the constant or the $data, run this test case with the original `private static ?Parser $parser;` will get error, but `private static ?Parser $parser = null;` not.

Ocramius replied:
So the triggering factor is `ReflectionFunctionAbstract::CLOSURE_NAME;`?

Asking because I'm mostly trying to reduce this further into a smaller test case, rather than a side-effect of code floating within the test suite.

Ocramius replied:
@lanlin looking at this again, can the test be formulated in form of something that actually calls `ReflectionFunctionAbstract`?

Does this test crash also when `@runInSeparateProcess` is used?

Ocramius commented:
Closing here: please re-open once there's a failing test case that shows an explicit failure, not by transitivity
